### PR TITLE
fix(ui/rag): UAT follow-ups — recompose capture guards, lazy-dep sweep, CPU-hang verdict (tasks 636-638)

### PR DESCRIPTION
**Pull request by rmusser01**
Three UAT follow-ups from the #899 stream:

- **task-636** (docs-only): CPU-hang investigation closed as an honest negative — not reproduced in 4 fresh sessions on current dev (incl. stress variants targeting the recompose race); stack skeleton mapped to Python 3.12's eager_task_factory (Textual sets it unconditionally); well-evidenced hypothesis that the task-627 capture fix removed the trigger cascade; websocket-latency vector explicitly documented as untested-not-ruled-out.
- **task-637**: new `RecomposeCaptureGuard` mixin generalizes the BaseAppScreen capture protection to 5 non-screen recompose widgets, with correctly narrowed scoping (releases only self-or-descendant captures — a sibling's legitimate capture survives, test-locked; verified against Textual 8.2.7 internals by the reviewer). ~22 further same-class sites enumerated and filed as task-660; a found-but-unrelated Chatbooks mount-before-attach crash filed as task-661.
- **task-638**: single `lazy_embeddings_rag_available()` seam in optional_deps now serves the search window's two reads AND EmbeddingFactory (no more stale-flag banners for users who have the deps); no-extras test coverage restored via per-test skipif; two more test files with the same anti-pattern fixed.

Review: all three Approved (one gate-count transcription error in 637's notes corrected per reviewer re-run). Gates: 71 search/embeddings/optional-deps, 578/8 RAG, capture suites green.

Held for owner approval.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Hardened mouse capture during non-screen recomposes and made Search/RAG dependency checks lazy and off the UI thread; `_start_indexing_run` now uses a fast spec probe. CPU-hang (task 636) is closed as not reproduced with findings documented.

- **Bug Fixes**
  - Added `RecomposeCaptureGuard` and applied it to `MCPRail`, `ChatbooksWindowImproved`, `MindmapViewerWindow`, `MediaIngestWindowRebuilt`, and `ResultsDashboardWindow`; releases only self-or-descendant captures; switched logging to `logger.opt(exception=True)`; added focused tests (includes deferred-teardown and sibling-capture cases).
  - Introduced `optional_deps.lazy_embeddings_rag_available()`; `Embeddings_Lib` and `SearchRAGWindow` now use it. The mount-time deep probe runs on a `@work(thread=True)` worker and applies via `call_from_thread`; `_start_indexing_run` uses `embeddings_rag_deps_installed()` to avoid blocking. Tests updated to wait for the worker, assert the deep probe isn’t called on button press, and restore no-extras coverage with per-test skips.

<sup>Written for commit 9953ec4159cd189956ec9fe78bb0f367e9cb1b85. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/rmusser01/tldw_chatbook/pull/905?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: tldw_chatbook/Widgets/recompose_capture_guard.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        repaint: bool = True,
+        layout: bool = False,
+        recompose: bool = False,
+    ) -> "Self":
+        """Recompose, releasing this widget's own stale mouse capture first.
+
+        Mirrors ``BaseAppScreen.refresh`` -- released here (at the moment
+        ``refresh(recompose=True)`` is CALLED) so the common, already-idle
+        case never depends on the deferred-teardown guard below at all.
+        """
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_chatbook/Widgets/recompose_capture_guard.py:120]: Fixed in cfa80aaa9: added Google-style `Args:`/`Returns:` sections to `RecomposeCaptureGuard.refresh()`'s docstring documenting `*regions`, `repaint`, `layout`, `recompose`, and the `self` return contract.

File: tldw_chatbook/Widgets/recompose_capture_guard.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+unconditional on attachment, same as ``BaseAppScreen``, since a detached
+widget can never be a legitimate capture for anyone.
+"""
+
+from typing import TYPE_CHECKING, Optional
+from loguru import logger
+
+from textual.geometry import Region
+from textual.widget import Widget
+
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_chatbook/Widgets/recompose_capture_guard.py:50]: Fixed in cfa80aaa9: merged the loguru and textual imports into a single contiguous third-party block (stdlib `typing` / third-party loguru+textual / local, no blank line splitting the third-party group).

File: tldw_chatbook/Widgets/recompose_capture_guard.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        captured = getattr(app, "mouse_captured", None)
+        if not self._capture_is_within_self(captured):
+            return
+        try:
+            app.capture_mouse(None)
+        except Exception:
+            logger.debug(
+                f"{type(self).__name__}: mouse-capture release {context} skipped.",
+                exc_info=True,
+            )
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_chatbook/Widgets/recompose_capture_guard.py:95]: Confirmed real (verified with a direct loguru-sink probe: exc_info=True produced no traceback text, opt(exception=True) did). Fixed in cfa80aaa9: both occurrences in this file now use logger.opt(exception=True).debug(...) instead of logger.debug(..., exc_info=True). Added Tests/Widgets/test_recompose_capture_guard.py::test_release_own_capture_failure_logs_traceback_via_loguru_opt, which is RED against the pre-fix code (asserts the emitted line contains the exception type/message/`Traceback`) and GREEN after.

File: tldw_chatbook/UI/Views/RAGSearch/search_rag_window.py [hunk 4]
```diff
@@ -551,8 +551,13 @@ async def on_mount(self) -> None:
...
-        # Check if embeddings/RAG dependencies are available
-        if not DEPENDENCIES_AVAILABLE.get('embeddings_rag', False):
+        # Check if embeddings/RAG dependencies are available. Routed through
+        # the lazy-ensure seam (task-638) rather than reading the raw flag:
+        # under the default lazy dependency-checking mode nothing else calls
+        # check_embeddings_rag_deps() before this runs, so the raw flag
+        # would stay stuck at its pristine False default and show this
+        # banner even when the packages are genuinely importable.
+        if not lazy_embeddings_rag_available():
             from ....Utils.widget_helpers import alert_embeddings_not_available
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_chatbook/UI/Views/RAGSearch/search_rag_window.py]: Fixed in cfa80aaa9. on_mount now trusts an already-True DEPENDENCIES_AVAILABLE['embeddings_rag'] read instantly (cheap); on a False read it dispatches the deep check to a new @work(thread=True) worker (_probe_embeddings_rag_availability) and applies the missing-deps banner via call_from_thread on completion, so the heavy torch/transformers/chromadb/sentence_transformers import probe never runs on the UI event loop during mount. Added a regression test (test_mount_time_dependency_probe_runs_off_the_ui_thread) asserting the probe is invoked from a different OS thread than the test/UI thread -- RED against the pre-fix code, GREEN after.

Also checked _start_indexing_run per your note: it is NOT already on a worker -- it's called directly from a synchronous @on(Button.Pressed, '#start-indexing') handler, so its guard was equally exposed. Rather than keep it synchronous, I switched it to the cheap, import-free embeddings_rag_deps_installed() spec probe (same pattern semantic_indexing_available() already uses for the backfill worker's own gate), so it never blocks the UI thread either; a 'specs present but genuinely broken' edge case still fails safely one round-trip later inside the existing backfill worker's error handling. Covered by a new test (test_start_indexing_guard_never_runs_the_deep_import_probe) that fails the deep probe loudly if it's ever called from this path -- RED against the pre-fix code, GREEN after.
